One-page Slack tutorial (#6698)
circleci/circleci-docs: jekyll/_cci2/slack-orb-tutorial.adoc

---
version:
- Cloud
- Server v3.x
- Server v2.x
---
= Slack Orb Tutorial
:page-layout: classic-docs
:page-liquid:
:page-description: "Learn how to get started with the CircleCI Slack orb"
:icons: font
:toc: macro
:toc-title:

This guide gets you started with the CircleCI Slack orb.

toc::[]

== Introduction

Orbs are open source, shareable packages of reusable configuration elements. They help you reduce complexity, integrate your software with apps, and save time.

One of CircleCI's most popular orbs is Slack. The Slack orb allows you to implement event-based notifications across all your CI/CD pipelines. Using either built-in message templates or Slack's visual Block Kit Builder, you can create and customize notifications specific to your organization's needs.

In this tutorial, you will learn how to:

> * Create a config.yml file using the Slack orb.

Reviewer: ```suggestion
* Create a `.circleci/config.yml` file using the Slack orb.
```

> * Store your Slack credentials securely in a Context.

Reviewer: ```suggestion
* Store your Slack credentials securely in a context.
```

                      "emoji": true
                    }
                  ]
                }
              ]
            }
          event: always
workflows:
  send-notification:
    jobs:
      - notify:
          context: slack-secrets
----

The following commentary describes what occurs in the main elements of the sample code:

* Line 1: This indicates the version of the CircleCI platform you are using. `2.1` is the most recent version.
* Lines 2-3: This is the orbs stanza. You are using the `slack: circleci/slack@4.9.3` orb.
* Lines 5-7: This is a job called _notify_, which is executed in a basic Docker container.
* Lines 8-9: This is a step which must be called _slack/notify_ to work with the Slack orb.
* Lines 10-24: This custom block includes a basic notification from the Slack Block Kit Builder. As you can see, it is in JSON format. In Part 3, you will use more sophisticated templates.
* Line 25: The `event: always` parameter means this notification is triggered for _all_ event types. In Part 3, you will use different notifications for different events.
* Line 30: This references a CircleCI context where you will save your Slack credentials in Step 2.

> Now you have a basic config.yml file set up, you can connect CircleCI with Slack.

Reviewer: ```suggestion
Now you have a basic `config.yml` file set up, you can connect CircleCI with Slack.
```

. On the _OAuth & Permissions_ page, scroll down to **Scopes**. This is where you need to create the permissions for your Slack app.
. Under **Bot Token Scopes**, click **Add an OAuth Scope**.
. The Slack orb needs permission to post chat messages and upload files, so create the following scopes:
* chat:write
* chat:write.public
* files:write

TIP: To receive Slack notifications in a private channel, youl will need to add your Slack app to that channel.

==== Install your app

. Once you have created your scopes, click the **Install to Workspace** button.
. You will then be asked to grant permission for the app to access your Slack workspace.
. Click the disclosure triangle to double-check the permissions, then click the green **Allow** button.
. You should see a *Bot User OAuth Token*. Copy this token to your clipboard, ready to add it to CircleCI. Make sure you keep this private.

=== Creating a context

In CircleCI, contexts allow you to secure and share environment variables across projects. Once you've created a context with your Slack credentials, you and your colleagues will be able to reuse them.

In CircleCI:

. Click the _Organization Settings_ page.

> . Click the blue *Create Context* button and add a unique name, such as _slack-secrets_ (that is the name specified in the config.yml file above).

Reviewer: ```suggestion
. Click the blue *Create Context* button and add a unique name, such as _slack-secrets_ (that is the name specified in the `config.yml` file above).
```

. Click the blue **Create Context** button.

> . Need to add something about security here.

Reviewer: Need security or delete point 😄

. Click the name of the context you just created.
. Click the blue *Add Environment Variable* button and enter your first key value pair. 
* The Environment Variable Name is SLACK_ACCESS_TOKEN.
* The value is your Slack Bot User OAuth Access Token.
. Click the Add Environment Variable button to save it.
. Click the blue **Add Environment Variable** button again.
* The Environment Variable Name is SLACK_DEFAULT_CHANNEL.
* The value is the ID of the default Slack channel for posting your notifications. You can override this setting in your individual jobs.

TIP: To get the ID for your Slack channel, right-click the channel in Slack and choose **Copy Link**. The ID will be visible at the end of the URL and will be in this format: C034R26AM36.

Make sure you have included the slack-secrets context in your _notify_ job and that the name matches:

> // screenshot of stanza

Reviewer: Needs screen shot

You can now reuse this context in other jobs and projects.

> Commit your config.yml file (and push it, if you are working remotely).

Reviewer: ```suggestion
Commit your `config.ym`l file (and push it, if you are working remotely).
```

==== Triggering an alert

In the CircleCI dashboard:

* Click **Projects**. 
* Find the repo and click the blue **Set Up Project** button next to it.

> * Choose the branch on which you committed your config.yml file

Reviewer: ```suggestion
* Choose the branch on which you committed your `config.yml` file
```

* Click the blue **Set Up Project** button.

This triggers your CircleCI pipeline, which contains a Slack orb with your credentials.

You should then see a green **Success** badge and a green tick next to your _notify_ job. 

Click on your job to see what just happened. You should see the message body that was sent to Slack.

Now open your Slack workspace. In the default channel you specified earlier, you should see the alert triggered by your CircleCI pipeline.

Although this is a basic alert, you have achieved a lot already:

> * Created a config.yml with the Slack orb.

Reviewer: ```suggestion
* Created a `.circleci/config.yml` with the Slack orb.
```
